OpenSim/Simulation/MarkersReference.cpp

```diff
@@ -22,139 +22,157 @@
-    Units modelUnits) : MarkersReference()
-{
+                                   Units modelUnits) :
+    MarkersReference() {
     loadMarkersFile(markerFile, modelUnits);
 }
 
-
-MarkersReference::MarkersReference(MarkerData *markerData,
-    const Set<MarkerWeight>* markerWeightSet) : MarkersReference()
-{
-    _markerData.reset(markerData);
-    if (markerWeightSet!=nullptr) upd_marker_weights()= *markerWeightSet;
-    populateFromMarkerData(*markerData);
+MarkersReference::
+MarkersReference(const TimeSeriesTable_<SimTK::Vec3>& markerTable,
+                 const Set<MarkerWeight>* markerWeightSet,
+                 Units units) :
+    MarkersReference() {
+    _markerTable = markerTable;
+    if(markerWeightSet != nullptr)
+        upd_marker_weights() = *markerWeightSet;
+    populateFromMarkerData(markerTable, units.getAbbreviation());
 }
 
-/* load the marker data for this MarkersReference from markerFile  */
-void MarkersReference::loadMarkersFile(const std::string markerFile, Units modelUnits)
-{
-    upd_marker_file() = markerFile;
-    _markerData.reset(new MarkerData(get_marker_file()));
+void MarkersReference::loadMarkersFile(const std::string markerFile,
+                                       Units modelUnits) {
+    auto fileExt = FileAdapter::findExtension(markerFile);
+    OPENSIM_THROW_IF(!(fileExt == "sto" || fileExt == "trc"),
+                     UnsupportedFileType,
+                     markerFile,
+                     "Supported file types are -- STO, TRC.");
+
+    if(fileExt == "trc") {
+        _markerTable = TimeSeriesTableVec3{markerFile};
+    } else {
+        try {
+            _markerTable = (TimeSeriesTable{markerFile}).pack<SimTK::Vec3>();
+        } catch(const IncorrectTableType&) {
+            _markerTable = TimeSeriesTable_<SimTK::Vec3>{markerFile};
+        }
+    }
 
-    // Convert the marker data into the model's units
-    _markerData->convertToUnits(modelUnits);
+    upd_marker_file() = markerFile;
 
-    populateFromMarkerData(*_markerData);
+    populateFromMarkerData(_markerTable, modelUnits.getAbbreviation());
 }
 
+void
+MarkersReference::
+populateFromMarkerData(const TimeSeriesTable_<SimTK::Vec3>& markerTable,
+                       const std::string& units) {
+    Units thisUnits{};
+    if(_markerTable.hasTableMetaDataKey("Units"))
+        thisUnits = Units{_markerTable.getTableMetaData<std::string>("Units")};
+    else
+        thisUnits = Units{Units::Meters};
+
+    double scaleFactor = thisUnits.convertTo(Units{units});
 
-/* A convenience method to populate MarkersReference from MarkerData **/
-void MarkersReference::populateFromMarkerData(const MarkerData& aMarkerData)
-{
-    const Array<std::string> &tempNames = aMarkerData.getMarkerNames();
-    int nm = tempNames.getSize();
+    OPENSIM_THROW_IF(SimTK::isNaN(scaleFactor),
+                     Exception,
+                     "Model has unspecified units.");
```

Comment:
The unit conversion code is `DataTable` analogue of `MarkerData::convertToUnits` which throws a similar exception. So I copied the same here.
I have changed the message to say `TimeSeriesTable has unspecified units.`.